ui/app/routes/application.js

```diff
@@ -1,16 +1,68 @@
 import { inject as service } from '@ember/service';
+import { next } from '@ember/runloop';
 import Route from '@ember/routing/route';
 import { AbortError } from 'ember-data/adapters/errors';
+import RSVP from 'rsvp';
 
 export default Route.extend({
   config: service(),
+  system: service(),
+  store: service(),
+
+  queryParams: {
+    region: {
+      refreshModel: true,
+    },
+  },
 
   resetController(controller, isExiting) {
     if (isExiting) {
       controller.set('error', null);
     }
   },
 
+  beforeModel(transition) {
+    return RSVP.all([this.get('system.regions'), this.get('system.defaultRegion')]).then(
+      promises => {
+        if (!this.get('system.shouldShowRegions')) return promises;
+
+        const queryParam = transition.queryParams.region;
+        const defaultRegion = this.get('system.defaultRegion.region');
+        const currentRegion = this.get('system.activeRegion') || defaultRegion;
+
+        // Only reset the store if the region actually changed
+        if (
+          (queryParam && queryParam !== currentRegion) ||
+          (!queryParam && currentRegion !== defaultRegion)
+        ) {
+          this.get('system').reset();
+          this.get('store').unloadAll();
+        }
+
+        this.set('system.activeRegion', queryParam || defaultRegion);
+
+        return promises;
+      }
+    );
+  },
+
+  // setupController doesn't refire when the model hook refires as part of
+  // a query param change
```

Comment:
😳 

Is that because the model hook is missing from this route? My understanding of the rule was that `setupController` fires on entering the route, and when `model` returns a truthy value.

If you had a trivial model hook, maybe that would cause `setupController` to fire?

```js
model(params) {
  return params.region;
},
```

Reply:
Innnnterresting. I'll give it a try.